### ISSUE TITLE
[Common Lisp] Implement new Concept Exercise: hash-tables

### ISSUE THREAD

**Issue by TheLostLambda** [exercism/v3#1819]
This issue describes how to implement the `hash-tables` concept exercise for the Common Lisp track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Learning objectives

- Understand what a hash-table is and how it can be used as map between keys and values
- Grasp the basic performance characteristics of hash-tables – O(1) insertion / lookup
- Be able to make a new hash-table with `make-hash-table`
- Be able to set the `:test` function of the hash-table and know `eql` is the default
- Know how to get a value from a hash-table using `gethash`
- Understand that `gethash` returns two values and that the second can be used to check if a key was present or not
- Be able to add an entry to a hash-table using `setf` and `gethash`
- Be able to remove and entry from a hash-table with `remhash`
- Know how to get the size of a hash-table with `hash-table-count`
- Be capable of hash-table iteration using `maphash`

## Out of scope

- The implementation details of hash-tables
- The exceptions to the O(1) rule (hash-table clashes)
- The more "unusual" methods of hash-table iteration (`with-hash-table-iterator` and `loop`)
- The "lower-level" options to `make-hash-table` (`:rehash-size` and `:rehash-threshold`)
- Association or property lists, those have their own exercises

## Concepts

- `hash-tables`

## Prerequisites

- `assignment`
- `multiple-values`
- `sameness`
- `symbols`
- `variables`
- `higher-order-functions`

## Resources to refer to

- [Old CL Cookbook](http://cl-cookbook.sourceforge.net/hashes.html)

### Hints

- [Tutorials Point](https://www.tutorialspoint.com/lisp/lisp_hash_table.htm)

### After

- [Performance Section of CL Cookbook](https://lispcookbook.github.io/cl-cookbook/data-structures.html#performance-issues-the-size-of-your-hash-table)

## Representer

Nothing yet

## Analyzer

Be sure that the student is using hash-tables and not alists or plists (though this can also likely be achieved via the test-suite)

## Implementing

Hash-tables (and "mapping" types in general) are very common in programming, so there should be lots you can do here. There are a lot of more ergonomic ways to represent maps in CL (like alists or plists), so it should be emphasised that hash-tables should be reserved for "high-performance" applications. The simpler list-based types might even be faster for small sizes.

Be sure the student knows to not overuse hash-tables.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.
